File: codeiq/tyama_codeiq711_next.lua
#!/usr/bin/lua
reverse=function(a,start,size)
	local end_=start+size-1
	for i=0,size/2-1 do
		z=a[start+i]
		a[start+i]=a[end_-i]
		a[end_-i]=z
	end
end
next_permutation=function(a,n)
	if not n then n=#a end
	if n<0 or #a<n then
		return false
	end
	local i=0
	reverse(a,n,#a+1-n)
	i=#a-1
	while i>=1 do
		if a[i]<a[i+1] then break end
		i=i-1
	end
	if i<1 then
		reverse(a,1,#a)
		return false
	end
	local k=i
	i=#a
	while i>=k+1 do
		if a[k]<a[i] then break end
		i=i-1
	end
	local l=i
	local z=a[k]
	a[k]=a[l]
	a[l]=z
	reverse(a,k+1,#a+1-(k+1))
	return true
end

N=6
e0={}
f0={}
r=0
for i=1,N do
	e0[i]=0
	f0[i]=0
	e0[N+i]=1
	f0[N+i]=1
end
table.sort(e0)
table.sort(f0)
repeat
	repeat
		flg=0
		zero1=0
		zero2=N
		one1=0
		one2=N
		for i=1,N*2 do
			if e0[i]==0 then zero1=zero1+1 end
			if e0[i]==1 then one1=one1+1 end
			if f0[N*2+1-i]==0 then zero2=zero2-1 end
			if f0[N*2+1-i]==1 then one2=one2-1 end
			if zero1==zero2 then flg=flg+1 end
			if one1==one2 then flg=flg+1 end
		end
		if flg>=2 then r=r+1 end
	until not next_permutation(f0)
until not next_permutation(e0)
print(r)

File: codeiq/tyama_codeiq711_next.go
//usr/bin/env go run $0 $@;exit
package main
import (
	"fmt"
	"sort"
)

func reverse(a sort.Interface,start int,size int){
	for end:=start+size-1;start<end;start++ {
		a.Swap(start,end)
		end--
	}
}
func next_permutation(a sort.Interface, n int) bool{
	if n<0||a.Len()<n {return false}
	reverse(a,n,a.Len()-n)
	i:=0
	for i=a.Len()-2;i>=0;i-- {if a.Less(i,i+1) {break}}
	if i<0 {
		reverse(a,0,a.Len())
		return false
	}
	k:=i
	for i=a.Len()-1;i>=k+1;i-- {if a.Less(k,i) {break}}
	l:=i
	a.Swap(k,l)
	reverse(a,k+1,a.Len()-(k+1))
	return true
}

func main(){
	N:=6
	e0:=make([]int,N*2)
	f0:=make([]int,N*2)
	i:=0
	r:=0
	for i=0;i<N;i++ {e0[N+i]=1;f0[N+i]=1}
	sort.Sort(sort.IntSlice(e0))
	sort.Sort(sort.IntSlice(f0))
	for {
		for {
			flg:=0
			zero1:=0
			zero2:=N
			one1:=0
			one2:=N
			for i:=0;i<N*2;i++ {
				if e0[i]==0 {zero1 += 1}
				if e0[i]==1 {one1  += 1}
				if f0[i]==0 {zero2 -= 1}
				if f0[i]==1 {one2  -= 1}
				if zero1==zero2 {flg+=1}
				if one1==one2 {flg+=1}
			}
			if flg>=2 {r+=1}
			if !next_permutation(sort.IntSlice(f0),len(f0)) {break}
		}
		if !next_permutation(sort.IntSlice(e0),len(e0)) {break}
	}
	fmt.Println(r)
}